sqlflow_models/cluster_keras.py

```diff
@@ -0,0 +1,272 @@
+
+        self.decoder_layers.append(Dense(units=self._pretrain_dims[0],
+                                         kernel_initializer=self._pretrain_initializer,
+                                         name='decoder_0'))
+        # Pretrain - autoencoder, encoder
+        self._autoencoder = keras.Sequential(layers=[self.input_layer] + self.encoder_layers + self.decoder_layers,
+                                             name='autoencoder')
+        self._autoencoder.compile(optimizer=self._pretrain_optimizer, loss='mse')
+        self._encoder = keras.Sequential(layers=[self.input_layer] + self.encoder_layers, name='encoder')
+        self._encoder.compile(optimizer=self._pretrain_optimizer, loss='mse')
+
+        if self._run_pretrain:
+            if self._existed_pretrain_model:
+                try:
+                    self._autoencoder.load_weights(filepath=self._existed_pretrain_model)
+                except FileNotFoundError as fe:
+                    print(fe)
+                    pass
+                except ImportError as ie:
+                    print(ie)
+                    pass
+            else:
+                raise ValueError('Model path should be specified when using pre-train model.')
+
+        else:
+            self._autoencoder.fit(x, x, batch_size=self._pretrain_batch_size, epochs=self._pretrain_epochs)
+
+        self.encoded_input = self._encoder.predict(x)
+
+        del self._autoencoder
+        del self._encoder
+        del self.decoder_layers
+
+    def call(self, inputs, training=None, mask=None):
+        x = self.input_layer(inputs)
+        for encoder_layer in self.encoder_layers:
+            x = encoder_layer(x)
+        return self.clustering_layer(x)
+
+    def init_centroids(self):
+        """
+        Training K-means `_kmeans_init` times on the output of encoder to get best initial centroids.
+        :return:
+        """
+        self.kmeans = KMeans(n_clusters=self._n_clusters, n_init=self._kmeans_init)
+        self.kmeans.fit_predict(self.encoded_input)
+
+    def evaluate(self, x=None, y=None, batch_size=None, verbose=1, sample_weight=None, steps=None,
+                 callbacks=None, max_queue_size=10, workers=1, use_multiprocessing=False):
+        metric = dict()
+        q = self.predict(x)
+        y_pred = q.argmax(1)
+        metric['acc'] = np.round(acc(y, y_pred), 5)
+        metric['nmi'] = np.round(nmi(y, y_pred), 5)
+        metric['ari'] = np.round(ari(y, y_pred), 5)
+        return q, metric
+
+    def cluster_train_loop(self, x, y, batch_size=256, maxiter=8000, update_interval=150, tol=0.001):
+        index_array = np.arange(x.shape[0])
+        # initialize centroids for clustering.
+        self.init_centroids()
+        self.get_layer(name='clustering').set_weights([self.kmeans.cluster_centers_])
+        self.display_model_info()
+        index, loss, p = 0, 0., None
+        y_pred_last = self.kmeans.fit_predict(self.encoded_input)
```

Comment:
This line is redundancy,  L175 can return the y_pred_last.

Reply:
Got it. Will remove it.